ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -696,15 +697,23 @@ impl<'a, B: ShardBinding> DZKPValidator for SemiHonestDZKPValidator<'a, B> {
     }
 }
 
+type DzkpBatcher<'a> = Batcher<'a, Batch>;
+
+/// The DZKP validator, and all associated contexts, each hold a reference to a single
+/// instance of `MaliciousDZKPValidatorInner`.
+pub(super) struct MaliciousDZKPValidatorInner<'a> {
+    pub(super) batcher: Mutex<DzkpBatcher<'a>>,
+    pub(super) validate_ctx: Base<'a>,
```

Comment:
We recently moved the `validate_record` function from the validator to the context. This was because `validate_record` gets called many times, but we wanted the validators to be unclonable so that the ownership is clear and there's a clearer driver of the "did everything get validated" check than dropping of the last `Arc`/`Weak` reference. However, `validate_record` needs a context to use for validation. It can't get it from the validator, because it doesn't have a reference to the validator, only to the batch.

From the `validate_record` change, up until this PR, the `validate_record` function was obtaining a context by narrowing the base context with the `DZKPValidate` step. That has the problem that we can't use the same `DZKPValidate` step for every instantiation of the malicious protocol. So I moved the validation context into the new `ValidatorInner` struct, which all of the contexts have a reference to.

An alternate solution would be to put the validation context inline in `UpgradedMaliciousContext`. But that gets replicated many times throughout execution of the protocol, so it seemed unnecessarily costly.

Reply:
Makes sense, thanks for clarifying!